### PR TITLE
interfaces: add an interface that allows clients to use media-hub over dbus

### PR DESCRIPTION
**Pull request by jhodapp**
Tweak some of the apparmor rules for the media-hub interface test.

Condense the explicit list of media-hub Session interface down to a single entry.

### CODE REVIEW

File: interfaces/builtin/media_hub.go [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+    interface=org.freedesktop.DBus
+    member="{Request,Release}Name"
+    peer=(name=org.freedesktop.DBus, label=unconfined),
+
+dbus (send)
+    bus=session
+    path=/org/freedesktop/DBus
+    interface=org.freedesktop.DBus
+    member="{Request,Release}Name"
+    peer=(name=org.freedesktop.DBus, label=unconfined),
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: Why are you registering on both the system and the session bus? All your dbus policy is focused on the session bus so maybe just remove the bind on the system bus?

File: interfaces/builtin/media_hub.go [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+    bus=system
+    path=/org/freedesktop/DBus
+    interface=org.freedesktop.DBus
+    member="GetConnectionUnix{ProcessID,User}"
+    peer=(label=unconfined),
+
+# Allow binding the service to the requested connection name
+dbus (bind)
+	bus=session
+	name="org.mpris.MediaPlayer2.MediaHub",
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: Please use the 'mpris' interface for this.

File: interfaces/builtin/media_hub.go [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+	bus=session
+	path=/com/ubuntu/media/Service{,/**}
+	interface=org.freedesktop.DBus**
+	peer=(label=unconfined),
+
+dbus (send)
+	bus=session
+	path=/org/freedesktop/Telepathy/AccountManager
+	interface=org.freedesktop.DBus.Properties
+	member="GetAll",
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: This seems wholly unrelated. Why is it needed? This doesn't exist in core so only another snap would be able to provide this (unless this is only needed on classic). Either way, this feels like another interface.

File: interfaces/builtin/media_hub.go [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+dbus (receive)
+	bus=session
+	path=/core/ubuntu/media/Service
+	peer=(label=@{profile_name}),
+
+# Allow all access to powerd for now, but we can fine-tune this if needed
+dbus (receive, send)
+	bus=system
+	path=/com/canonical/powerd
+	interface=com.canonical.powerd,
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: This is also unrelated. Perhaps this is part of the unity8-session interface?

File: interfaces/builtin/media_hub.go [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+
+dbus (receive, send)
+	bus=system
+	path=/com/canonical/Unity/Screen
+	interface=com.canonical.Unity.Screen,
+`)
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: Please have your snap `plugs: [ screen-inhibit-control]` instead of including this here.

File: interfaces/builtin/media_hub.go [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+    path=/core/ubuntu/media/Service
+    peer=(label=###PLUG_SECURITY_TAGS###),
+
+# Allow clients to manager Player sessions
+dbus (receive)
+    bus=session
+    path=/core/ubuntu/media/Service
+    interface=core.ubuntu.media.Service
+    member="{Create,Detach,Reattach,Destroy,CreateFixed,Resume}Session"
+	peer=(label=###PLUG_SECURITY_TAGS###),
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: Can you fix the whitespace here?

File: interfaces/builtin/media_hub.go [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+
+# Allow clients to pause all other sessions
+dbus (receive)
+    bus=session
+    path=/core/ubuntu/media/Service
+    interface=core.ubuntu.media.Service
+    member="PauseOtherSessions"
+	peer=(label=###PLUG_SECURITY_TAGS###),
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: and here?

File: interfaces/builtin/media_hub.go [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+
+# Allow clients to query/modify service properties
+dbus (receive)
+    bus=session
+    path=/core/ubuntu/media/Service
+    interface=org.freedesktop.DBus.Properties
+    member="{Get,Set}"
+    peer=(label=###PLUG_SECURITY_TAGS###),
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: Is 'Set()' considered safe for all clients? I suppose so-- the apparmor-easyprof-ubuntu policy allowed access to the entire DBus API. Is there a reason you aren't simply doing this instead of all the fine-grained policy (this of course is assuming the entire API is safe):
```
dbus (receive)
    bus=session
    path=/core/ubuntu/media/Service{,/**}
    peer=(label=###PLUG_SECURITY_TAGS###),
```

File: interfaces/builtin/media_hub.go [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+dbus (receive)
+    bus=session
+    path=/core/ubuntu/media/Service
+    peer=(label=unconfined),
+
+dbus (receive)
+    bus=session
+    interface=org.freedesktop.DBus.Introspectable
+    peer=(label=unconfined),
+`)
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: This should probably just be moved to mediaHubPermanentSlotAppArmor since it offers no security benefit on all snaps (unconfined is free to change apparmor profiles and is considered trusted).

File: interfaces/builtin/media_hub.go [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+
+var mediaHubConnectedPlugAppArmor = []byte(`
+# Description: Allow using media-hub service. Reserved because this gives
+#  privileged access to the service.
+# Usage: reserved
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: Please update the description. media-hub client access is not privileged and is considered safe. Please drop 'Usage' altogether.

File: interfaces/builtin/media_hub.go [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+import (
+	"bytes"
+
+	"github.com/snapcore/snapd/interfaces"
+)
+
+var mediaHubPermanentSlotAppArmor = []byte(`
+# Description: Allow operating as the the media-hub service. Reserved because
+#  this gives privileged access to the system.
+# Usage: reserved
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: I'm surprised more access to the system isn't required. I guess you are using 'plugs: [ camera ]`? There was a lot of hybris and accesses in /dev specific to android drivers in the debian packaging apparmor profile... I guess you are skipping all of that for now.

File: interfaces/builtin/media_hub.go [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+    interface=org.freedesktop.DBus.Properties
+    member="{Get,Set}"
+    peer=(label=###SLOT_SECURITY_TAGS###),
+
+dbus (receive)
+   bus=system
+   path=/core/ubuntu/media/Service
+   interface=org.freedesktop.DBus.Properties
+   member=PropertiesChanged
+   peer=(label=###SLOT_SECURITY_TAGS###),
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: Can all of the above be simplified as:
```
# Allow communications with media-hub
dbus (receive, send)
     bus=session
     path=/core/ubuntu/media/Service{,/**}
     peer=(label=###SLOT_SECURITY_TAGS###),
```

File: interfaces/builtin/media_hub.go [hunk 13]
```diff
@@ -0,0 +1,301 @@
...
+
+dbus (receive)
+    bus=system
+    path=/
+    interface=org.freedesktop.DBus.ObjectManager
+    peer=(label=unconfined),
+`)
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: This is curious. Why is it needed?

File: interfaces/builtin/media_hub.go [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+
+var mediaHubPermanentSlotSecComp = []byte(`
+getsockname
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: This is in the default policy.

File: interfaces/builtin/media_hub.go [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
+recvmsg
+sendmsg
+sendto
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: If you have sendto, I suspect you will need `recvfrom`. Feel free to add it.

File: interfaces/builtin/media_hub.go [hunk 16]
```diff
@@ -0,0 +1,301 @@
...
+`)
+
+var mediaHubConnectedPlugSecComp = []byte(`
+getsockname
+recvmsg
+sendmsg
+sendto
+`)
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: Same here.

File: interfaces/builtin/media_hub.go [hunk 17]
```diff
@@ -0,0 +1,301 @@
...
+    <allow send_interface="core.ubuntu.media.Service"/>
+</policy>
+`)
+
+var mediaHubConnectedPlugDBus = []byte(`
+<policy context="default">
+    <deny own="core.ubuntu.media.Service"/>
+    <allow send_destination="core.ubuntu.media.Service"/>
+    <allow send_interface="core.ubuntu.media.Service"/>
+</policy>
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: Can you explain this policy? I suspect you are hitting up against the fact that Ubuntu Core doesn't yet support user sessions. You'll want to talk to ted and tvoss about the status of that work.

File: interfaces/builtin/media_hub.go [hunk 18]
```diff
@@ -0,0 +1,301 @@
...
+func (iface *MediaHubInterface) PermanentPlugSnippet(plug *interfaces.Plug, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	return nil, nil
+}
+
+func (iface *MediaHubInterface) ConnectedPlugSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityAppArmor:
+		old := []byte("###SLOT_SECURITY_TAGS###")
+		new := slotAppLabelExpr(slot)
+		snippet := bytes.Replace(mediaHubConnectedPlugAppArmor, old, new, -1)
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: What are your use cases for this interface? All snaps? classic + media-hub snap? classic + media-hub deb? All of them?

If classic + media-hub deb is one of them, I suggest adjusted this section and make the value of `new` conditional on release.OnClassic such that when on classic `new` is `unconfined`, otherwise it is slotAppLabelExpr(slot). See the modem-manager interface for an example.

File: interfaces/builtin/media_hub.go [hunk 19]
```diff
@@ -0,0 +1,301 @@
...
+
+func (iface *MediaHubInterface) ConnectedSlotSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityAppArmor:
+		old := []byte("###PLUG_SECURITY_TAGS###")
+		new := plugAppLabelExpr(plug)
+		snippet := bytes.Replace(mediaHubConnectedSlotAppArmor, old, new, -1)
+		return snippet, nil
+	}
+	return nil, nil
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: As mentioned above, please drop this and just put the mediaHubConnectedSlotAppArmor policy in mediaHubPermanentSlotAppArmor.

**Comment by chipaca** [comment]
Closing because the author of the PR has not responded to feedback in over two months.

**Comment by jhodapp** [comment]
We're still around and actively working on this at the moment. Please re-open.

File: interfaces/builtin/media_hub.go (hunk 1 above)
Comment (jhodapp) [inline interfaces/builtin/media_hub.go]: Yeah indeed, this just needs to be cleaned up.

File: interfaces/builtin/media_hub.go (hunk 11 above)
Comment (jhodapp) [inline interfaces/builtin/media_hub.go]: Yes, your assessment is correct. First pass is just for audio playback with the more complex system access required by video to come second (it's actually in progress now).

File: interfaces/builtin/media_hub.go (hunk 2 above)
Comment (jhodapp) [inline interfaces/builtin/media_hub.go]: Unfortunately the mpris interface doesn't cut it here as media-hub does things both on two different interface names, one being media-hub specific like shown here. This is an artifact from the very first bits of media-hub code and I'm not exactly sure why it is using both.

File: interfaces/builtin/media_hub.go (hunk 3 above)
Comment (jhodapp) [inline interfaces/builtin/media_hub.go]: Just my lack of understanding of interfaces at the time I wrote this.

File: interfaces/builtin/media_hub.go (hunk 4 above)
Comment (jhodapp) [inline interfaces/builtin/media_hub.go]: See last comment about interface understanding.

File: interfaces/builtin/media_hub.go (hunk 13 above)
Comment (jhodapp) [inline interfaces/builtin/media_hub.go]: Started from location-service as an example. Probably just left over from that.

File: interfaces/builtin/media_hub.go (hunk 14 above)
Comment (jhodapp) [inline interfaces/builtin/media_hub.go]: Where is this default policy? I didn't even realize it would inherit anything.

File: interfaces/builtin/media_hub.go (hunk 17 above)
Comment (jhodapp) [inline interfaces/builtin/media_hub.go]: Again, used location-service as a base.

File: interfaces/builtin/media_hub.go (hunk 18 above)
Comment (jhodapp) [inline interfaces/builtin/media_hub.go]: Use case would be all 3 of those. I don't exactly understand what classic implies for apparmor policy yet. Any good reading pointers for that?

File: interfaces/builtin/media_hub.go (hunk 17 above)
Comment (jhodapp) [inline interfaces/builtin/media_hub.go]: Update after talking with Simon: Basically since we're including media-hub as part of the unity8-session snap, nothing is blocking this access anyway for the permanentSlot/connectedPlug so there's no need to explicitly allow with these rules, so let's remove these two.

File: interfaces/builtin/media_hub.go (hunk 13 above)
Comment (bergotorino) [inline interfaces/builtin/media_hub.go]: This is removed

File: interfaces/builtin/media_hub.go (hunk 9 above)
Comment (bergotorino) [inline interfaces/builtin/media_hub.go]: Moved

File: interfaces/builtin/media_hub.go (hunk 1 above)
Comment (bergotorino) [inline interfaces/builtin/media_hub.go]: Addressed

File: interfaces/builtin/media_hub.go (hunk 19 above)
Comment (bergotorino) [inline interfaces/builtin/media_hub.go]: Dropped

File: interfaces/builtin/basedeclaration.go [hunk 20]
```diff
@@ -326,6 +326,11 @@ slots:
       slot-snap-type:
         - core
     deny-auto-connection: true
+  media-hub:
+    allow-installation:
+      slot-snap-type:
+        - app
+    deny-auto-connection: true
```
Comment (jdstrand) [inline interfaces/builtin/basedeclaration.go]: We allowed using media-hub on Touch and I expected us to allow this interface to be auto-connected. Why did you specify `deny-auto-connection: true` here?

File: interfaces/builtin/media_hub.go [hunk 21]
```diff
@@ -0,0 +1,218 @@
...
+package builtin
+
+import (
+	"bytes"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/release"
+)
+
+var mediaHubPermanentSlotAppArmor = []byte(`
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: Please adjust to use:
```
const mediaHubPermanentSlotAppArmor = `
...
```

While there may be a few older interfaces that use `var`, using `const` is the coding style for all new interfaces.

File: interfaces/builtin/media_hub.go [hunk 22]
```diff
@@ -0,0 +1,218 @@
...
+# Description: Allow operating as the the media-hub service. Reserved because
+# this gives privileged access to the system.
+# Usage: reserved
+
+# DBus accesses
+#include <abstractions/dbus-strict>
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: Your policy below is for the session bus, yet you are using the system bus abstraction. Please adjust to be:
```
#include <abstractions/dbus-session-strict>
```

File: interfaces/builtin/media_hub.go [hunk 23]
```diff
@@ -0,0 +1,218 @@
...
+	"bytes"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/release"
+)
+
+var mediaHubPermanentSlotAppArmor = []byte(`
+# Description: Allow operating as the the media-hub service. Reserved because
+# this gives privileged access to the system.
+# Usage: reserved
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: Please remove the 'Usage' meta tag. It is no longer needed with the base declaration.

File: interfaces/builtin/media_hub.go [hunk 24]
```diff
@@ -0,0 +1,218 @@
...
+import (
+	"bytes"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/release"
+)
+
+var mediaHubPermanentSlotAppArmor = []byte(`
+# Description: Allow operating as the the media-hub service. Reserved because
+# this gives privileged access to the system.
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: This interface doesn't provide privileged access to the system, as such, this comment should be changed to:
```
# Description: Allow operating as the the media-hub trusted helper service.
```

File: interfaces/builtin/media_hub.go (hunk 2 above)
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: @jhodapp - I'm saying use the mpris interface for name="org.mpris.MediaPlayer2.MediaHub" instead of reimplementing that policy here and keep the name="core.ubuntu.media.Service" that is defined below.

File: interfaces/builtin/media_hub.go [hunk 25]
```diff
@@ -0,0 +1,218 @@
...
+dbus (receive)
+	bus=session
+	path=/core/ubuntu/media/Service
+	peer=(label=@{profile_name}),
+
+dbus (receive)
+    bus=session
+    interface=org.freedesktop.DBus.Introspectable
+    peer=(label=unconfined),
+
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: The policy above this comment is fine for mediaHubPermanentSlotAppArmor, but the policy below this comment should be moved to mediaHubConnectedSlotAppArmor.

File: interfaces/builtin/media_hub.go [hunk 26]
```diff
@@ -0,0 +1,218 @@
...
+# Allow binding the service to the requested connection name
+dbus (bind)
+	bus=session
+	name="core.ubuntu.media.Service",
+
+dbus (receive, send)
+	bus=session
+	path=/com/ubuntu/media/Service{,/**}
+	interface=org.freedesktop.DBus**
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: This should be: `interface=org.freedesktop.DBus{,.*}`

File: interfaces/builtin/media_hub.go [hunk 27]
```diff
@@ -0,0 +1,218 @@
...
+    member=PropertiesChanged
+    peer=(label=###PLUG_SECURITY_TAGS###),
+`)
+
+var mediaHubConnectedSlotAppArmorClassic = []byte(`
+# Allow unconfined clients to interact with the player on classic
+dbus (receive)
+    bus=session
+    path=/core/ubuntu/media/Service
+    peer=(label=unconfined),
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: This rule is a superset of this rule in mediaHubPermanentSlotAppArmor:
```
dbus (receive, send)
	bus=session
	path=/com/ubuntu/media/Service{,/**}
	interface=org.freedesktop.DBus**
	peer=(label=unconfined),
```

It looks like you are not using this variable anywhere. Please remove.

File: interfaces/builtin/media_hub.go [hunk 28]
```diff
@@ -0,0 +1,218 @@
...
+`)
+
+var mediaHubConnectedPlugAppArmor = []byte(`
+# Description: Allow using media-hub service.
+
+#include <abstractions/dbus-strict>
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: Please use: `#include <abstractions/dbus-session-strict>`

File: interfaces/builtin/media_hub.go [hunk 29]
```diff
@@ -0,0 +1,218 @@
...
+dbus (receive, send)
+    bus=session
+    path=/core/ubuntu/media/Service{,/**}
+    peer=(label=###SLOT_SECURITY_TAGS###),
+
+dbus (receive)
+    bus=system
+    path=/
+    interface=org.freedesktop.DBus.ObjectManager
+    peer=(label=unconfined),
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: Why is this needed?

File: interfaces/builtin/media_hub.go [hunk 30]
```diff
@@ -0,0 +1,218 @@
...
+
+var mediaHubConnectedSlotAppArmorClassic = []byte(`
+# Allow unconfined clients to interact with the player on classic
+dbus (receive)
+    bus=session
+    path=/core/ubuntu/media/Service
+    peer=(label=unconfined),
+`)
+
+var mediaHubConnectedPlugAppArmor = []byte(`
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: Please use `const`

File: interfaces/builtin/media_hub.go [hunk 31]
```diff
@@ -0,0 +1,218 @@
...
+		return mediaHubPermanentSlotAppArmor, nil
+	case interfaces.SecuritySecComp:
+		return mediaHubPermanentSlotSecComp, nil
+	}
+	return nil, nil
+}
+
+func (iface *MediaHubInterface) ConnectedSlotSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	return nil, nil
+}
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: ConnectedSlotSnippet should do the same thing for `###PLUG_SECURITY_TAGS###` and ConnectedSlotAppArmor that ConnectedPlugSnippet does for `###SLOT_SECURITY_TAGS###"` and ConnectedPlugAppArmor.

In other words, you need to define in ConnectedSlotSnippet the policy needed for the server to receive the plugs side  communication just like ConnectedPlugSnippet defines the policy needed for the client to send the slots side communication to the server.

File: interfaces/builtin/media_hub.go [hunk 32]
```diff
@@ -0,0 +1,218 @@
...
+	return nil
+}
+
+func (iface *MediaHubInterface) SanitizeSlot(slot *interfaces.Slot) error {
+	return nil
+}
+
+func (iface *MediaHubInterface) LegacyAutoConnect() bool {
+	return false
+}
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: Please remove this.

File: interfaces/builtin/basedeclaration.go (hunk 20 above)
Comment (bergotorino) [inline interfaces/builtin/basedeclaration.go]: I'm not sure actually, this has been there since the beginning nevertheless I'll remove it.

File: interfaces/builtin/media_hub.go (hunk 21 above)
Comment (bergotorino) [inline interfaces/builtin/media_hub.go]: Done

File: interfaces/builtin/media_hub.go (hunk 29 above)
Comment (bergotorino) [inline interfaces/builtin/media_hub.go]: @jhodapp do you know the origin of the above?

File: interfaces/builtin/basedeclaration.go [hunk 33]
```diff
@@ -326,6 +326,10 @@ slots:
       slot-snap-type:
         - core
     deny-auto-connection: true
+  media-hub:
+    allow-installation:
+      slot-snap-type:
+        - app
```
Comment (jdstrand) [inline interfaces/builtin/basedeclaration.go]: Thanks for removing the deny-auto-connection. I forgot to mention that since this is not an implicit slot, we need a deny-connection constraint. It also occurred to me that you are supporting talking to the deb on classic so this needs to be adjusted to be:
```
  media-hub:
    allow-installation:
      slot-snap-type:
        - app
        - core
    deny-connection:
      on-classic: false
```

This is saying that on core, allow connections to the snap and on all-snaps a snap declaration is required to connect to the snap. When connection is allowed, allow auto-connection.

File: interfaces/builtin/media_hub.go [hunk 34]
```diff
@@ -0,0 +1,219 @@
...
+
+dbus (receive)
+    bus=session
+    interface=org.freedesktop.DBus.Introspectable
+    peer=(label=unconfined),
+`
+
+const mediaHubConnectedSlotAppArmor = `
+# Allow connected clients to interact with the service
+
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: This comment seems redundant considering the comment just below.

File: interfaces/builtin/media_hub.go [hunk 35]
```diff
@@ -0,0 +1,219 @@
...
+	switch securitySystem {
+	case interfaces.SecurityAppArmor:
+		old := []byte("###PLUG_SECURITY_TAGS###")
+		new := []byte("")
+		if release.OnClassic {
+			// On classic MediaHub will run unconfined
+			new = []byte("unconfined")
+		} else {
+			new = plugAppLabelExpr(plug)
+		}
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: Do you expect that when OnClassic, media-hub will always be the deb or can it be a snap? If only a deb, you can drop the OnClassic stuff because it is already unconfined. When a snap, you actually don't want either/or for the label, you always want the `plugAppLabelExpr(plug)`, but OnClassic you want additional policy for unconfined. However, since only allowing unconfined to talk to media-hub OnClassic doesn't provide a security benefit, please change this function to:
```
func (iface *MediaHubInterface) ConnectedSlotSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
	switch securitySystem {
	case interfaces.SecurityAppArmor:
		old := []byte("###PLUG_SECURITY_TAGS###")
                new = plugAppLabelExpr(plug)
                snippet := bytes.Replace([]byte(mediaHubConnectedSlotAppArmor), old, new, -1)
                return snippet, nil
        }
        return nil, nil
}
```

and change mediaHubPermanentSlotAppArmor to include:
```
# Allow unconfined to talk to us
dbus (receive, send)
    bus=session
    path=/core/ubuntu/media/Service{,/**}
    peer=(label=unconfined),
```

File: interfaces/builtin/media_hub.go (hunk 2 above)
Comment (jhodapp) [inline interfaces/builtin/media_hub.go]: @jdstrand Either there is a typo in your reply or I don't understand what you mean yet by "instead of reimplementing that policy here and keep the name="core.ubuntu.media.Service" that is defined below." Do you mean to not keep the dbus (bind) portion below that uses name="core.ubuntu.media.Service" and only keep the one above that uses "org.mpris.MediaPlayer2.MediaHub"?

**Comment by jdstrand** [comment]
@jhodapp - I have a comment in my inbox but don't see it here. It seems there is some confusion with what I said here:

"
I'm saying use the mpris interface for name="org.mpris.MediaPlayer2.MediaHub" instead of reimplementing that policy here and keep the name="core.ubuntu.media.Service"[sic] that is defined below.
"

I'm saying that we have a mpris interface so remove your dbus rules for org.mpris.MediaPlayer2.MediaHub from this PR. Leave all the other rules for com.ubuntu.media.Service in this PR like we've been iterating on. Then your implementation can do in its snapcraft.yaml:

```
name: media-hub
slots:
  session-service:
    interface: media-hub
  mpris-service:
    interface: mpris
    name: "MediaHub"
...
```

Put another way, the media-hub interface should implement media-hub-y things and the mpris interface should implement mpris-y things. Your implementation happens to be able to do both, so you slots both.

File: interfaces/builtin/media_hub.go [hunk 36]
```diff
@@ -0,0 +1,203 @@
...
+    bus=session
+    path=/com/ubuntu/media/Service{,/**}
+    interface=org.freedesktop.DBus{,.*}
+    peer=(label=unconfined),
+
+# We can always connect to ourselves
+dbus (receive)
+    bus=session
+    path=/core/ubuntu/media/Service
+    peer=(label=@{profile_name}),
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: Note with https://github.com/snapcore/snapd/pull/2768, this rule can be dropped.

File: interfaces/builtin/media_hub.go [hunk 37]
```diff
@@ -0,0 +1,203 @@
...
+    interface=org.freedesktop.DBus.Introspectable
+    peer=(label=###PLUG_SECURITY_TAGS###),
+
+dbus (receive)
+    bus=session
+    interface="core.ubuntu.media.Service{,.*}"
+    path=/core/ubuntu/media/Service
+    peer=(label=###PLUG_SECURITY_TAGS###),
+
+# Allow clients to manager Player sessions
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: Type: 'manager' should be 'manage'

File: interfaces/builtin/media_hub.go [hunk 38]
```diff
@@ -0,0 +1,203 @@
...
+
+func (iface *MediaHubInterface) ConnectedPlugSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityAppArmor:
+		old := []byte("###SLOT_SECURITY_TAGS###")
+		new := slotAppLabelExpr(slot)
+		if release.OnClassic {
+			// On classic MediaHub will run unconfined
+			new = []byte("unconfined")
+		}
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: You didn't quite follow my advice in https://github.com/snapcore/snapd/pull/2230#discussion_r98787150. You added the recommended rule to mediaHubPermanentSlotAppArmor but you forgot to remove the `if release.OnClassic` logic.

File: interfaces/builtin/media_hub_test.go [hunk 39]
```diff
@@ -0,0 +1,189 @@
...
+
+	snippet, err := s.iface.ConnectedPlugSnippet(s.plug, s.slot, system)
+	c.Assert(err, IsNil)
+	c.Assert(snippet, Not(IsNil))
+
+	c.Assert(string(snippet), testutil.Contains, "#include <abstractions/dbus-session-strict>")
+	c.Assert(string(snippet), testutil.Contains, "peer=(label=unconfined),")
+}
+
+func (s *MediaHubInterfaceSuite) TestPermanentPlugSnippetAppArmor(c *C) {
```
Comment (jdstrand) [inline interfaces/builtin/media_hub_test.go]: This should be named TestPermanentSlotSnippetAppArmor

**Comment by bergotorino** [comment]
@jdstrand 

I'm still parsing your comment trying to understand and provide an answer. Apologies it takes time. Anyways, just to clarify I'm on a right track:

What do you mean by "rules to querying libapparmor", should I understand it like "/lib/@{multiarch}/libapparmor.so* mr" entry?

**Comment by bergotorino** [comment]
@jdstrand pushed the rule to allow querying libapparmor

File: interfaces/builtin/media_hub.go [hunk 40]
```diff
@@ -0,0 +1,200 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: It's 2017 now.

File: interfaces/builtin/media_hub.go [hunk 41]
```diff
@@ -0,0 +1,200 @@
...
+    path=/org/freedesktop/DBus
+    interface=org.freedesktop.DBus
+    member="{Request,Release}Name"
+    peer=(name=org.freedesktop.DBus, label=unconfined),
+
+# Allow binding the service to the requested connection name
+dbus (bind)
+    bus=session
+    name="core.ubuntu.media.Service",
+
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: Please add a comment here: `# Allow communications with unconfined processes`

File: interfaces/builtin/media_hub.go [hunk 42]
```diff
@@ -0,0 +1,200 @@
...
+dbus (receive, send)
+    bus=session
+    path=/com/ubuntu/media/Service{,/**}
+    interface=org.freedesktop.DBus{,.*}
+    peer=(label=unconfined),
+
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: Please add: `# Allow unconfined processes to introspect us`

File: interfaces/builtin/media_hub.go [hunk 43]
```diff
@@ -0,0 +1,200 @@
...
+dbus (receive)
+    bus=session
+    interface=org.freedesktop.DBus.Introspectable
+    peer=(label=unconfined),
+
+dbus (receive, send)
+    bus=session
+    path=/core/ubuntu/media/Service{,/**}
+    peer=(label=unconfined),
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: This rule is a superset of the rule above that uses "interface=org.freedesktop.DBus{,.*}". Can we limit this to a particular DBus interface? Perhaps `interface="core.ubuntu.media.Service{,.*}"` like in the seemingly corresponding rule in mediaHubConnectedSlotAppArmor for snap connections?

File: interfaces/builtin/media_hub.go [hunk 44]
```diff
@@ -0,0 +1,200 @@
...
+    path=/core/ubuntu/media/Service
+    interface=core.ubuntu.media.Service
+    member="PauseOtherSessions"
+    peer=(label=###PLUG_SECURITY_TAGS###),
+
+# Allow clients to query/modify service properties
+dbus (receive)
+    bus=session
+    path=/core/ubuntu/media/Service{,/**}
+    peer=(label=###PLUG_SECURITY_TAGS###),
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: This is a superset of every receive rule that uses this path. Can you add a path= or member= clause to this rule?

File: interfaces/builtin/media_hub.go [hunk 45]
```diff
@@ -0,0 +1,200 @@
...
+
+const mediaHubConnectedPlugAppArmor = `
+# Description: Allow using media-hub service.
+
+#include <abstractions/dbus-session-strict>
+
+dbus (receive, send)
+    bus=session
+    path=/core/ubuntu/media/Service{,/**}
+    peer=(label=###SLOT_SECURITY_TAGS###),
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: This is a very open rule, but the ConnectedSlot rules are more fine-grained. If it is safe for the client to talk to the service on all methods, feel free to adjust the ConnectedSlot policy to use:
```
dbus (receive, send)
￼    bus=session
    path=/core/ubuntu/media/Service{,/**}
    peer=(label=###PLUG_SECURITY_TAGS###),
```

If there are some APIs that should not be accessible to plugs, please list the safe ones in both ConnectedSlot and ConnectedPlug.

File: interfaces/builtin/media_hub.go [hunk 46]
```diff
@@ -0,0 +1,200 @@
...
+sendto
+recvfrom
+`
+
+const mediaHubConnectedPlugSecComp = `
+recvmsg
+sendmsg
+sendto
+recvfrom
+`
```
Comment (jdstrand) [inline interfaces/builtin/media_hub.go]: You can drop mediaHubPermanentSlotSecComp and mediaHubConnectedPlugSecComp now, these syscalls are all available by default now.

File: interfaces/builtin/media_hub_test.go [hunk 47]
```diff
@@ -0,0 +1,188 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
```
Comment (jdstrand) [inline interfaces/builtin/media_hub_test.go]: It's 2017 now.

**Comment by niemeyer** [comment]
@jhodapp @bergotorino Review above is 16 days old, 24 days since it was last touched.

Any feedback here?

**Comment by bergotorino** [comment]
@niemeyer this is still on, all review comments will be addressed. stay tuned.

**Comment by alfonsosanchezbeato** [comment]
Please close this one and follow on the review in https://github.com/snapcore/snapd/pull/3048

**Comment by jhodapp** [comment]
Closing at Alfonso's request. See comment above for where the current PR lives.
